### PR TITLE
Fix XmlSerializer Serializing DateTime when DataType = "time"

### LINKED ISSUE #6761: [Port]  XmlSerializer incorrectly serializing Fields with [XmlText(DataType = "time")]
XmlSerializer incorrectly serializes DateTime fields/properties marked with  [XmlText(DataType = "time")] if the DateTime value is of UTC kind.

### PR DESCRIPTION
**Pull request by shmao**
XmlSerializer incorrectly serializes DateTime fields/properties marked with [XmlText(DataType = "time")] if the DateTime value is of UTC kind. The UTC DateTime was serialized as a local DateTime and wouldn't make a round trip.

Fix #6761

### CODE REVIEW

**Comment by shmao** [comment]
cc: @SGuyGe @khdang

**Comment by shmao** [comment]
@dotnet-bot test Innerloop OSX Release Build and Test please

File: src/System.Xml.XmlSerializer/tests/XmlSerializerTests.cs [hunk 1]
```diff
@@ -101,6 +101,34 @@ public static void Xml_DateTimeAsRoot()
...
+        DateTime localTime = new DateTime(549269870000L, DateTimeKind.Local);
+        DateTimeDataTypeTimeType localTimeOjbect = new DateTimeDataTypeTimeType()
+        {
+            Value = localTime
+        };
+
+        string localTimeString = localTime.ToString("HH:mm:ss.fffffffzzzzzz");
+        DateTimeDataTypeTimeType localTimeOjbectRoundTrip = SerializeAndDeserialize(localTimeOjbect,
+string.Format(@"<?xml version=""1.0"" encoding=""utf-8""?>
+<DateTimeDataTypeTimeType xmlns:xsi=""http://www.w3.org/2001/XMLSchema-instance"" xmlns:xsd=""http://www.w3.org/2001/XMLSchema"">{0}</DateTimeDataTypeTimeType>", localTimeString));
```
Comment (jasonwilliams200OK) [inline src/System.Xml.XmlSerializer/tests/XmlSerializerTests.cs]: Nit: C#6's `verbatim string interpolation`:

```c#
        DateTimeDataTypeTimeType localTimeOjbectRoundTrip = SerializeAndDeserialize(localTimeOjbect,
$@"<?xml version=""1.0"" encoding=""utf-8""?>
<DateTimeDataTypeTimeType xmlns:xsi=""http://www.w3.org/2001/XMLSchema-instance"" xmlns:xsd=""http://www.w3.org/2001/XMLSchema"">{localTimeString}</DateTimeDataTypeTimeType>");
```

File: src/System.Xml.XmlSerializer/src/System/Xml/Serialization/XmlCustomFormatter.cs [hunk 2]
```diff
@@ -28,7 +28,8 @@ internal static string FromDate(DateTime value)
 
         internal static string FromTime(DateTime value)
         {
-            return XmlConvert.ToString(DateTime.MinValue + value.TimeOfDay, "HH:mm:ss.fffffffzzzzzz");
+            string dateFormat = value.Kind == DateTimeKind.Utc ? "HH:mm:ss.fffffffZ" : "HH:mm:ss.fffffffzzzzzz";
```
Comment (SGuyGe) [inline src/System.Xml.XmlSerializer/src/System/Xml/Serialization/XmlCustomFormatter.cs]: Here can we simply use the standard format "o" that covers all cases?

**Comment by khdang** [comment]
LGTM

File: src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs [hunk 3]
```diff
@@ -1285,6 +1285,18 @@ public void Add(DateTime value)
         }
     }
 
+    public class DateTimeDataTypeTimeType
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs]: It's hard to tell what this type is about, what about TypeWithDateTimePropertyAsXmlTime?

File: src/System.Xml.XmlSerializer/tests/XmlSerializerTests.cs [hunk 4]
```diff
@@ -101,6 +101,34 @@ public static void Xml_DateTimeAsRoot()
     }
 
     [Fact]
+    public static void Xml_DateTime_DataType_Time()
```
Comment (SGuyGe) [inline src/System.Xml.XmlSerializer/tests/XmlSerializerTests.cs]: Remove extra '_' inside the type name.

**Comment by SGuyGe** [comment]
A few comments, otherwise LGTM.

File: src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs (hunk 3 above)
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs]: Changed the name per the suggestion. Thanks.

File: src/System.Xml.XmlSerializer/src/System/Xml/Serialization/XmlCustomFormatter.cs (hunk 2 above)
Comment (shmao) [inline src/System.Xml.XmlSerializer/src/System/Xml/Serialization/XmlCustomFormatter.cs]: To keep the same behavior as desktop I'd prefer keep this as is.

**Comment by shmao** [comment]
I changed the fix per the feedbacks. But the test is currently broken probably due to the Daylight Saving Time change. There might be a bug in either the code or the test. I'm investigating this.

**Comment by shmao** [comment]
The new added test is failing due to #7012.

**Comment by shmao** [comment]
@SGuyGe @khdang @zhenlan Can you please have a look this [additional change](https://github.com/dotnet/corefx/pull/6762/files#diff-5bae44dca35b891b75bf93d4e05d1efdR33)? Thanks.

**Comment by zhenlan** [comment]
LGTM.

@shmao can you please merge before Sunday fork?

**Comment by shmao** [comment]
@dotnet-bot test this please

**Comment by shmao** [comment]
@dotnet-bot test this please

**Comment by shmao** [comment]
@dotnet-bot test Innerloop Windows_NT Release Build and Test please

**Pull request merged**